syspagegen/syspagegen.c

```diff
@@ -0,0 +1,822 @@
+		do {
+			prog = CAST_SYSPTR(syspage_prog_t *, ptr);
+			printf("\t\t%s\n", CAST_SYSPTR(char *, prog->argv));
+		} while ((ptr = prog->next) != sysgen_common.syspage->progs);
+	}
+	else {
+		printf("\t\tnot defined\n");
+	}
+}
+
+
+static void sysgen_help(const char *prog)
+{
+	printf("Usage: %s [options] to create binary syspage file or add syspage to image\n", prog);
+	printf("\t-s <pimg> <offs> <sz>   - syspage properties\n");
+	printf("\t    pimg - begining physical address of the target image\n");
+	printf("\t    offs - syspage's offset in the target image\n");
+	printf("\t    sz   - max syspage's size\n");
+	printf("\t-p <path>               - path to preinit script\n");
+	printf("\t-u <path>               - path to user script\n");
+	printf("\t-b <path>               - path to output syspage binary file\n");
+	printf("\t-i <path>               - path to image \n");
+	printf("\t-h                      - print help message\n");
+}
+
+
+int main(int argc, char *argv[])
+{
+	int opt, err;
+	char *endptr;
+	const char *preinitScript, *userScript, *binName, *imgName;
+
+	preinitScript = userScript = binName = imgName = NULL;
+
+	if (argc <= 1) {
+		sysgen_help(argv[0]);
+		return EXIT_FAILURE;
+	}
+
+	while ((opt = getopt(argc, argv, "s:p:u:b:i:h")) != -1) {
+		switch (opt) {
+			case 's':
+				if ((optind + 1) < argc) {
+					sysgen_common.pkernel = strtoul(argv[optind - 1], &endptr, 0);
+					if ((sysgen_common.pkernel == 0 && strlen(endptr) != 0) || (sysgen_common.pkernel != 0 && strlen(endptr) != 0)) {
+						fprintf(stderr, "Wrong physical image address %s\n", argv[optind - 1]);
+						return -1;
+					}
+
+					sysgen_common.offs = strtoul(argv[optind], &endptr, 0);
+					if ((sysgen_common.offs == 0 && strlen(endptr) != 0) || (sysgen_common.offs != 0 && strlen(endptr) != 0)) {
+						fprintf(stderr, "Wrong syspage offset %s\n", argv[optind]);
+						return -1;
+					}
+
+					sysgen_common.maxsz = strtoul(argv[optind + 1], &endptr, 0);
+					if ((sysgen_common.maxsz == 0 && strlen(endptr) != 0) || (sysgen_common.maxsz != 0 && strlen(endptr) != 0)) {
+						fprintf(stderr, "Wrong syspage size %s\n", argv[optind + 1]);
+						return -1;
+					}
+				}
+				break;
+
+			case 'p':
+				preinitScript = optarg;
+				break;
+
+			case 'u':
+				userScript = optarg;
+				break;
+
+			case 'b':
+				binName = optarg;
+				break;
+
+			case 'i':
+				imgName = optarg;
+				break;
+
+			case 'h':
+			default:
+				sysgen_help(argv[0]);
+				return EXIT_SUCCESS;
+		}
+	}
+
+	sysgen_common.buff = calloc(sysgen_common.maxsz, sizeof(uint8_t));
+	if (sysgen_common.buff == NULL)
+		return EXIT_FAILURE;
+
+	sysgen_common.syspage = (syspage_t *)sysgen_common.buff;
+	sysgen_common.syspage->size = ALIGN_ADDR(sizeof(syspage_t), sizeof(long long));
+	sysgen_common.syspage->pkernel = sysgen_common.pkernel;
+	LIST_INIT(&sysgen_common.aliases);
+
+	/* Parse preinit script with map and console commands */
+	err = sysgen_parseScript(preinitScript);
```

Comment:
oh, so these params are actually not optional... so they should not be "options" (options per definition should be optional or mutually exclusive :)). You can probably keep it as-is, but specify in help message that it's mandatory.

Do we always need 2 scripts? or maybe we can support more than 2? Than we can maybe parse all remaining positional args as scripts?

Reply:
Frankly, I don't think that there will be a case for more than 2 scripts. The first one, `preinit script` contains basic information about platform like maps, devices etc. The second one consists of programs and aliases to them., that's all.

Ok, You have right it should not be defined as "options", I will change  it to "mandatory".